### DISCUSSION TITLE
Messaging: How do you read custom data?

### DISCUSSION THREAD

**Discussion by nijakobius** [capawesome-team/capacitor-firebase#421] [Q&A] [answered]
When  scheduling a Firebase Messaging campaign, you can optionally pass custom data to be read in the app (see attached screen shot).

Any idea how to read that data from within the app?

I've tried all of this plugin's functions and event listeners (like `notificationReceived`) and none of them worked. I've also tried [Capcitor's App plugin's](https://capacitorjs.com/docs/apis/app#getlaunchurl) functions and event listeners (like the `appUrlOpen`).

Or is there an alternative way for the app to know which it is opened from?

<img width="548" alt="Screenshot 2023-08-12 at 12 56 04" src="https://github.com/capawesome-team/capacitor-firebase/assets/37240226/02c4f6cf-4828-4a07-b5ec-045159b6d608">

**Comment by robingenz** [comment]
This is the UI for in-app messaging and not messaging. An in-app messaging plugin does not exist yet, see #129.

**Reply by DevDianDankie** [reply]
I am wondering is capacitor's push notifications plugin won't work? or are they not compatible?

**Reply by robingenz** [reply]
What do you mean exactly?

**Comment by nijakobius** [comment]
> This is the UI for in-app messaging and not messaging. An in-app messaging plugin does not exist yet, see https://github.com/capawesome-team/capacitor-firebase/issues/129.

No, this is the notifications UI. The one you get from Firebase > Messaging > New Campaign > Notifications (not In-app messaging) > Additional options (optional) > Custom data.

(Or am I misunderstanding what you meant?)

<img width="1639" alt="Screenshot 2023-08-12 at 22 04 24" src="https://github.com/capawesome-team/capacitor-firebase/assets/37240226/44599504-6b26-4661-818c-e9ca346bed90">

<img width="1674" alt="Screenshot 2023-08-12 at 22 08 09" src="https://github.com/capawesome-team/capacitor-firebase/assets/37240226/515bf9e0-d503-43ff-bc55-e9b67010299a">

**Reply by robingenz** [reply, marked as answer]
Okay, this should work. Have you tried the following?

```ts
import { FirebaseMessaging } from '@capacitor-firebase/messaging';

const addNotificationReceivedListener = async () => {
  await FirebaseMessaging.addListener('notificationReceived', event => {
    console.log(event.notification.data);
  });
};
```

**Reply by nijakobius** [reply]
That's it! I'm pretty sure I had tried that but maybe I did a mistake. This time round it worked - thanks! :)

**Reply by loama** [reply]
this listener isn't working for me on iOS (haven't tested it on Android), is there a bug on 5.1.0?

`"@capacitor-firebase/messaging": "^5.1.0"`

**Reply by loama** [reply]
Ok my issue was this one https://github.com/capawesome-team/capacitor-firebase/pull/267#issuecomment-1328885820
Should have read the docs with more attention
